pkg/networkservice/common/heal/client.go

```diff
@@ -19,41 +19,177 @@ package heal
-	cc  networkservice.MonitorConnectionClient
+	ctx      context.Context
+	cc       networkservice.MonitorConnectionClient
+	initOnce sync.Once
+	initErr  error
+	conns    connectionInfoMap
 }
 
-// NewClient - creates a new networkservice.NetworkServiceClient chain element that inform healServer about new client connection
+// NewClient - creates a new networkservice.NetworkServiceClient chain element
+//             - ctx - context for the lifecycle of the *Client* itself.  Cancel when discarding the client.
+//             - cc  - MonitorConnectionClient that will be used to watch for connection confirmations and breakages.
 func NewClient(ctx context.Context, cc networkservice.MonitorConnectionClient) networkservice.NetworkServiceClient {
 	return &healClient{
 		ctx: ctx,
 		cc:  cc,
 	}
 }
 
+func (u *healClient) init(ctx context.Context, conn *networkservice.Connection) error {
+	monitorClient, err := u.cc.MonitorConnections(u.ctx, &networkservice.MonitorScopeSelector{
+		PathSegments: []*networkservice.PathSegment{{Name: conn.GetCurrentPathSegment().Name}, {Name: ""}},
+	})
+	if err != nil {
+		return errors.Wrap(err, "MonitorConnections failed")
+	}
+
+	healConnection := requestHealConnectionFunc(ctx)
+	if healConnection == nil {
+		healConnection = func(conn *networkservice.Connection) {}
+	}
+	restoreConnection := requestRestoreConnectionFunc(ctx)
+	if restoreConnection == nil {
+		restoreConnection = func(conn *networkservice.Connection) {}
+	}
+
+	go u.listenToConnectionChanges(healConnection, restoreConnection, monitorClient)
+
+	return nil
+}
+
 func (u *healClient) Request(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (*networkservice.Connection, error) {
+	conn := request.GetConnection()
+	u.initOnce.Do(func() {
+		u.initErr = u.init(ctx, conn)
+	})
+	// if initialization failed, then we want for all subsequent calls to Request() on this object to also fail
+	if u.initErr != nil {
+		return nil, u.initErr
+	}
+
+	connInfo, _ := u.conns.LoadOrStore(conn.GetId(), &connectionInfo{
+		conn: conn.Clone(),
+		cond: sync.NewCond(&sync.Mutex{}),
+	})
+	u.replaceConnectionPath(conn, connInfo)
+
 	conn, err := next.Client(ctx).Request(ctx, request, opts...)
+	if err != nil {
+		return nil, err
+	}
 
-	registerClient := registerClientFunc(ctx)
-	if err == nil && registerClient != nil {
-		registerClient(u.ctx, conn, u.cc)
+	connInfo.cond.L.Lock()
+	defer connInfo.cond.L.Unlock()
+	connInfo.conn = conn.Clone()
+	if connInfo.state != connStateReady {
+		ch := make(chan struct{})
+		go func() {
+			connInfo.cond.Wait()
+			close(ch)
+		}()
+		select {
+		case <-ch:
+		case <-time.After(100 * time.Millisecond):
+			connInfo.cond.Broadcast()
+			<-ch
+			if connInfo.state == connStateInitial {
+				u.conns.Delete(conn.GetId())
+			}
+			_, _ = next.Client(ctx).Close(ctx, conn, opts...)
+			return nil, errors.Errorf("couldn't verify that connection: %v was established", conn.GetId())
```

Comment:
Finally, by my local investigation, I found that we are using the monitor because we incorrectly handling grpc connection death.